File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/ResultAssetSerializator.java
package com.greentree.engine.moon.assets.serializator;

import com.greentree.commons.reflection.info.TypeInfo;
import com.greentree.engine.moon.assets.asset.Asset;
import com.greentree.engine.moon.assets.asset.ConstAsset;
import com.greentree.engine.moon.assets.key.AssetKey;
import com.greentree.engine.moon.assets.key.ResultAssetKey;
import com.greentree.engine.moon.assets.serializator.manager.AssetManager;

public final class ResultAssetSerializator<T> extends TypedAssetSerializator<T> {

    public ResultAssetSerializator(TypeInfo<T> type) {
        super(type);
    }

    @Override
    public boolean canLoad(AssetManager manager, AssetKey key) {
        if (key instanceof ResultAssetKey res) {
            final var type = getType();
            return type.isInstance(res.result());
        }
        return false;
    }

    @SuppressWarnings("unchecked")
    @Override
    public Asset<T> load(AssetManager context, AssetKey key) {
        if (key instanceof ResultAssetKey res) {
            return new ConstAsset<>((T) res.result());
        }
        return null;
    }

}

File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/ResultAssetSerializator.kt
package com.greentree.engine.moon.assets.serializator

import com.greentree.commons.reflection.info.TypeInfo
import com.greentree.engine.moon.assets.asset.Asset
import com.greentree.engine.moon.assets.asset.ConstAsset
import com.greentree.engine.moon.assets.key.AssetKey
import com.greentree.engine.moon.assets.key.ResultAssetKey
import com.greentree.engine.moon.assets.serializator.manager.AssetManager
import com.greentree.engine.moon.assets.serializator.marker.NotMyKeyType

class ResultAssetSerializator<T : Any>(type: TypeInfo<T>) : TypedAssetSerializator<T>(type) {

	override fun load(context: AssetManager, key: AssetKey): Asset<T>? {
		if(key is ResultAssetKey) {
			val result = key.result
			if(type.isInstance(result))
				return ConstAsset(result as T)
			return null
		}
		throw NotMyKeyType
	}
}